File: android-design-patterns-and-best-practice/app3/app/src/main/java/edu/keepaneye/app3/Sandwich.java
package edu.keepaneye.app3;

import android.util.Log;

import java.util.ArrayList;
import java.util.List;

public class Sandwich {
    public static final String TAG = "tag";
    private List<Ingredient> ingredients = new ArrayList<>();

    public void getCalories() {
        int c = 0;
        for (Ingredient i : ingredients) {
            c += i.calories();
        }
        Log.d(TAG, "Total calories : " + c + " kcal");
    }

    public void addIngredient(Ingredient ingredient) {
        ingredients.add(ingredient);
    }

    public void getSandwich() {
        for (Ingredient i : ingredients) {
            Log.d(TAG, i.name() + " : " + i.calories() + " kcal");
        }
    }
}

File: android-design-patterns-and-best-practice/app7_1/app/src/main/java/edu/keepaneye/app7_1/builder/Sandwich.kt
package edu.keepaneye.app7_1.builder

class Sandwich {
    private val ingredients : MutableList<Ingredient> = mutableListOf()

    public fun addIngredient(i: Ingredient) {
        ingredients.add(i)
    }

    public fun getKcal(): Int {
        var kcal: Int = 0
        for(ingredient in ingredients) {
            kcal += ingredient.kcal()
        }
        return kcal
    }

    public fun getSandwich() : String {
        val sb: StringBuilder  =  StringBuilder()
        for(ingredient in ingredients) {
            sb.append(ingredient.description()).append("\n")
        }
        return sb.toString()
    }
}

class SandwichBuilder {
    fun build(sandwich: Sandwich, ingredient: Ingredient): Sandwich {
        sandwich.addIngredient(ingredient)
        return sandwich
    }
}